### ISSUE TITLE
Unexpected null ref

### ISSUE THREAD

**Issue by jakearchibald** [preactjs/preact#4049]
- [x] Check if updating to the latest Preact version resolves the issue

**Describe the bug**

A ref is unexpectedly null when using a particular JSX structure.

**To Reproduce**

[Demo](https://preactjs.com/repl?code=aW1wb3J0IHsgcmVuZGVyIH0gZnJvbSAicHJlYWN0IjsKaW1wb3J0IHsgdXNlU3RhdGUsIHVzZVJlZiwgdXNlRWZmZWN0IH0gZnJvbSAicHJlYWN0L2hvb2tzIjsKCmZ1bmN0aW9uIERlbW8oKSB7CiAgY29uc3QgW3BoYXNlLCBzZXRQaGFzZV0gPSB1c2VTdGF0ZSgxKTsKICBjb25zdCBvbmVSZWYgPSB1c2VSZWYobnVsbCk7CiAgY29uc3QgdHdvUmVmID0gdXNlUmVmKG51bGwpOwogIGNvbnN0IHRocmVlUmVmID0gdXNlUmVmKG51bGwpOwoKICB1c2VFZmZlY3QoKCkgPT4gewogICAgc2V0VGltZW91dCgoKSA9PiBzZXRQaGFzZSgyKSwgMTAwMCk7CiAgfSwgW10pOwoKICB1c2VFZmZlY3QoKCkgPT4gewogICAgY29uc29sZS5sb2coewoJCSAJcGhhc2UsCiAgICAgIG9uZVJlZjogb25lUmVmLmN1cnJlbnQsCiAgICAgIHR3b1JlZjogdHdvUmVmLmN1cnJlbnQsCiAgICAgIHRocmVlUmVmOiB0aHJlZVJlZi5jdXJyZW50LAogICAgfSk7CiAgfSk7CgogIHJldHVybiAoCiAgICA8PgogICAgICB7cGhhc2UgPT09IDEgPyAoCiAgICAgICAgPGRpdj4KICAgICAgICAgIDxkaXYgcmVmPXt0d29SZWZ9PkVsZW1lbnQgdHdvPC9kaXY%2BCiAgICAgICAgICA8ZGl2IHJlZj17dGhyZWVSZWZ9PkVsZW1lbnQgdGhyZWU8L2Rpdj4KICAgICAgICA8L2Rpdj4KICAgICAgKSA6IHBoYXNlID09PSAyID8gKAogICAgICAgIDxkaXYgY2xhc3M9Im91dGVyIj4KICAgICAgICAgIDxkaXYgcmVmPXtvbmVSZWZ9PkVsZW1lbnQgb25lPC9kaXY%2BCiAgICAgICAgICA8ZGl2IGNsYXNzPSJ3cmFwcGVyIj4KICAgICAgICAgICAgPGRpdiByZWY9e3R3b1JlZn0%2BRWxlbWVudCB0d288L2Rpdj4KICAgICAgICAgICAgPGRpdiByZWY9e3RocmVlUmVmfT5FbGVtZW50IHRocmVlPC9kaXY%2BCiAgICAgICAgICA8L2Rpdj4KICAgICAgICA8L2Rpdj4KICAgICAgKSA6IG51bGx9CiAgICA8Lz4KICApOwp9CgpyZW5kZXIoPERlbW8gLz4sIGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKCJhcHAiKSk7Cg%3D%3D).

View the console. I see:

```
{phase: 1, oneRef: null, twoRef: div, threeRef: div}
{phase: 2, oneRef: div, twoRef: null, threeRef: div}
```

**Expected behaviour**

```
{phase: 1, oneRef: null, twoRef: div, threeRef: div}
{phase: 2, oneRef: div, twoRef: div, threeRef: div}
```

It's expected that `oneRef` is `null` in phase 1 (the element isn't rendered), but `twoRef` shouldn't be `null` in phase 2, as the element _is_ rendered.

The bug seems to depend on a few things:

If `<div class="wrapper">` is switched to a fragment, the bug still happens. However, if `<div class="wrapper">` is removed, the bug does not happen.

If `<div class="outer">` is switched to a fragment, the bug does not happen.

**Comment by developit** [comment]
Ah, yeah. This is a bug in the ref "unset" code. When I switch to function refs, I can see the following ref updates:

```jsx
// phase 1
assigning twoRef <div>​Element two​</div>​
assigning threeRef <div>​Element three​</div>​

// phase 2
assigning twoRef <div>​Element two​</div>​
assigning threeRef <div>​Element three​</div>​
assigning twoRef null    // <--- this the bug
assigning oneRef <div>​Element one​</div>​
```

**Comment by jakearchibald** [comment]
I guess the unsetting code needs to check that it's unsetting the value it set.

**Comment by JoviDeCroock** [comment]
Hmm, I have investigated this issue before while working on v11. Basically the issue stems from the parent indentation being introduced. We recurse down the tree initially and see the following

```
<div> <-- depth 1
  <div ref={twoRef} <-- depth 2
  <div ref={threeRef} <-- depth 2
```

Notice the depth annotation as this will play a role here, when we go into the second phase we see the following


```
<div> <-- depth 1
  <div ref={oneRef} <-- depth 2
  <div <-- depth 2
    <div ref={twoRef} <-- depth 3
    <div ref={threeRef} <-- depth 3
```

So while we go through the tree we'll end up at depth 3 first to start performing operations, we see that `twoRef` and `threeRef` are "new" as there was nothing on depth 3 before so we create the elements and we assign the refs. We continue on to depth 2 where we notice a difference as we now miss `twoRef` and have a new `oneRef`, we'll create the `oneRef` VNode and unmount the `twoRef` VNode, this unmount leads us to call it with null to properly clean up the ref.

**Comment by developit** [comment]
Yeah I think this basically requires doing all ref updates in commit. To do that we need to expand the definition of what we store in the commit queue to include different functions (basically opcodes).

**Comment by JoviDeCroock** [comment]
The concern I have are layout and normal effects we need to guarantee refs being present 😅 I think storing it in the commit queue will indeed fix this

**Comment by jakearchibald** [comment]
Here's a little workaround in case people need it:

```js
import { useRef } from "preact/hooks";

const modifiedThisPass = new WeakSet();
const shittyProxies = new WeakMap();

export function shittyElRef() {
  const ref = useRef(null);

  if (!shittyProxies.has(ref)) {
    shittyProxies.set(ref, {
      get current() {
        return ref.current;
      },
      set current(val) {
        if (modifiedThisPass.has(ref)) {
          // If this has already been set this pass,
          // ignore null values
          if (val === null) return;
        } else {
          modifiedThisPass.add(ref);
          queueMicrotask(() => modifiedThisPass.delete(ref));
        }

        ref.current = val;
      },
    });
  }

  return shittyProxies.get(ref);
}
```

It tries to ensure that refs can't be set from a value to `null` in the same pass. [Demo](https://preactjs.com/repl?code=aW1wb3J0IHsgcmVuZGVyIH0gZnJvbSAicHJlYWN0IjsKaW1wb3J0IHsgdXNlU3RhdGUsIHVzZVJlZiwgdXNlRWZmZWN0IH0gZnJvbSAicHJlYWN0L2hvb2tzIjsKCmNvbnN0IG1vZGlmaWVkVGhpc1Bhc3MgPSBuZXcgV2Vha1NldCgpOwpjb25zdCBzaGl0dHlQcm94aWVzID0gbmV3IFdlYWtNYXAoKTsKCmZ1bmN0aW9uIHNoaXR0eUVsUmVmKCkgewogIGNvbnN0IHJlZiA9IHVzZVJlZihudWxsKTsKCiAgaWYgKCFzaGl0dHlQcm94aWVzLmhhcyhyZWYpKSB7CiAgICBzaGl0dHlQcm94aWVzLnNldChyZWYsIHsKICAgICAgZ2V0IGN1cnJlbnQoKSB7CiAgICAgICAgcmV0dXJuIHJlZi5jdXJyZW50OwogICAgICB9LAogICAgICBzZXQgY3VycmVudCh2YWwpIHsKICAgICAgICBpZiAobW9kaWZpZWRUaGlzUGFzcy5oYXMocmVmKSkgewogICAgICAgICAgLy8gSWYgdGhpcyBoYXMgYWxyZWFkeSBiZWVuIHNldCB0aGlzIHBhc3MsCiAgICAgICAgICAvLyBpZ25vcmUgbnVsbCB2YWx1ZXMKICAgICAgICAgIGlmICh2YWwgPT09IG51bGwpIHJldHVybjsKICAgICAgICB9IGVsc2UgewogICAgICAgICAgbW9kaWZpZWRUaGlzUGFzcy5hZGQocmVmKTsKICAgICAgICAgIHF1ZXVlTWljcm90YXNrKCgpID0%2BIG1vZGlmaWVkVGhpc1Bhc3MuZGVsZXRlKHJlZikpOwogICAgICAgIH0KCiAgICAgICAgcmVmLmN1cnJlbnQgPSB2YWwgfHwgbnVsbDsKICAgICAgfSwKICAgIH0pOwogIH0KCiAgcmV0dXJuIHNoaXR0eVByb3hpZXMuZ2V0KHJlZik7Cn0KCmZ1bmN0aW9uIERlbW8oKSB7CiAgY29uc3QgW3BoYXNlLCBzZXRQaGFzZV0gPSB1c2VTdGF0ZSgxKTsKICBjb25zdCBvbmVSZWYgPSBzaGl0dHlFbFJlZigpOwogIGNvbnN0IHR3b1JlZiA9IHNoaXR0eUVsUmVmKCk7CiAgY29uc3QgdGhyZWVSZWYgPSBzaGl0dHlFbFJlZigpOwoKICB1c2VFZmZlY3QoKCkgPT4gewogICAgc2V0VGltZW91dCgoKSA9PiBzZXRQaGFzZSgyKSwgMTAwMCk7CiAgICBzZXRUaW1lb3V0KCgpID0%2BIHNldFBoYXNlKDMpLCAyMDAwKTsKICB9LCBbXSk7CgogIHVzZUVmZmVjdCgoKSA9PiB7CiAgICBjb25zb2xlLmxvZyh7CiAgICAgIHBoYXNlLAogICAgICBvbmVSZWY6IG9uZVJlZi5jdXJyZW50LAogICAgICB0d29SZWY6IHR3b1JlZi5jdXJyZW50LAogICAgICB0aHJlZVJlZjogdGhyZWVSZWYuY3VycmVudCwKICAgIH0pOwogIH0pOwoKICBpZiAocGhhc2UgPT09IDMpIHJldHVybiA8ZGl2PjwvZGl2PjsKCiAgcmV0dXJuICgKICAgIDw%2BCiAgICAgIHtwaGFzZSA9PT0gMSA%2FICgKICAgICAgICA8ZGl2PgogICAgICAgICAgPGRpdiByZWY9e3R3b1JlZn0%2BRWxlbWVudCB0d288L2Rpdj4KICAgICAgICAgIDxkaXYgcmVmPXt0aHJlZVJlZn0%2BRWxlbWVudCB0aHJlZTwvZGl2PgogICAgICAgIDwvZGl2PgogICAgICApIDogcGhhc2UgPT09IDIgPyAoCiAgICAgICAgPGRpdiBjbGFzcz0ib3V0ZXIiPgogICAgICAgICAgPGRpdiByZWY9e29uZVJlZn0%2BRWxlbWVudCBvbmU8L2Rpdj4KICAgICAgICAgIDxkaXYgY2xhc3M9IndyYXBwZXIiPgogICAgICAgICAgICA8ZGl2IHJlZj17dHdvUmVmfT5FbGVtZW50IHR3bzwvZGl2PgogICAgICAgICAgICA8ZGl2IHJlZj17dGhyZWVSZWZ9PkVsZW1lbnQgdGhyZWU8L2Rpdj4KICAgICAgICAgIDwvZGl2PgogICAgICAgIDwvZGl2PgogICAgICApIDogbnVsbH0KICAgIDwvPgogICk7Cn0KCnJlbmRlcig8RGVtbyAvPiwgZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoImFwcCIpKTsK).

**Comment by JoviDeCroock** [comment]
@jakearchibald this should be fixed in 10.16.0, thank you so much again for the reproduction! It helped a ton in identifying this!

**Comment by jakearchibald** [comment]
Thanks for fixing it so quickly!
